### ISSUE TITLE
Support distributed training

### ISSUE THREAD

**Issue by EHxuban11** [Libre-YOLO/libreyolo#122]
A redditor has showed interest in distributed training: 

https://www.reddit.com/r/computervision/comments/1souw5j/comment/oh4tdwv/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

This is something that should be carefully planned, we can discuss it in this issue.

**Comment by imagra93** [comment]
# Multi-GPU Training

I have been checking on easy ways to implement this and we have (at least) two very distinct approaches that I would like to discuss.

---

## Approach 1 — Raw PyTorch DDP

### How it works

Each GPU runs a separate Python process. PyTorch's `DistributedDataParallel` (DDP)
synchronises gradients via an all-reduce after every backward pass. All processes run
the same training loop; only rank 0 saves checkpoints and runs validation.

### Launch

Replace `python train.py ...` with:

```bash
torchrun --nproc_per_node=4 train.py ...
```

`torchrun` sets `LOCAL_RANK`, `RANK`, and `WORLD_SIZE` in the environment.

---

### Changes — `libreyolo/training/trainer.py`

#### `__init__` — initialise the process group

Add at the top of `__init__`, before `_setup_device()`:

```python
import os
import torch.distributed as dist

self.rank       = int(os.environ.get("RANK", 0))
self.local_rank = int(os.environ.get("LOCAL_RANK", 0))
self.world_size = int(os.environ.get("WORLD_SIZE", 1))

if self.world_size > 1 and not dist.is_initialized():
    dist.init_process_group(backend="nccl")
```

Store `self.is_distributed = self.world_size > 1`.

#### `_setup_device` (line 156) — remove the `NotImplementedError`

The entire multi-GPU branch raises today. Replace with:

```python
def _setup_device(self) -> torch.device:
    if self.is_distributed:
        torch.cuda.set_device(self.local_rank)
        return torch.device(f"cuda:{self.local_rank}")
    # ... existing single-device logic unchanged ...
```

#### `effective_lr` property (line 88) — scale by world size

Each GPU receives `config.batch` samples; the global batch is
`config.batch * world_size`. The linear-scaling rule must account for this:

```python
@property
def effective_lr(self) -> float:
    return self.config.lr0 * self.config.batch * self.world_size / 64
```

Without this change every run with N GPUs would train with an LR that is N× too low.

#### `_setup_data` (line 231) — add `DistributedSampler`

```python
from torch.utils.data import DistributedSampler

sampler = (
    DistributedSampler(train_dataset, num_replicas=self.world_size,
                       rank=self.rank, shuffle=True)
    if self.is_distributed else None
)

self.train_loader = create_dataloader(
    train_dataset,
    batch_size=self.config.batch,
    num_workers=self.config.workers,
    shuffle=(sampler is None),
    pin_memory=True,
    sampler=sampler,
)
```

The sampler partitions the dataset across ranks. Setting `shuffle=False` is required
when a sampler is provided; the sampler itself handles shuffling.

Also call `sampler.set_epoch(epoch)` at the start of each epoch inside
`_train_epoch` so shuffling differs between epochs.

#### `setup` (line 339) — wrap model with DDP, gate I/O on rank 0

The order matters: the optimizer must be built from the **unwrapped** model, then DDP
is applied, then EMA is attached to `model.module`.

```python
def setup(self):
    if self._is_setup:
        return

    self.model.to(self.device)
    self.on_setup()           # accesses model.head etc. — must be before DDP
    self._setup_data()
    self.optimizer = self._setup_optimizer()   # also before DDP

    # Wrap AFTER optimizer so named_parameters() sees the raw model
    if self.is_distributed:
        self.model = nn.parallel.DistributedDataParallel(
            self.model, device_ids=[self.local_rank]
        )

    # ... scaler setup unchanged ...

    if self.config.ema:
        raw = self.model.module if self.is_distributed else self.model
        self.ema_model = ModelEMA(raw, decay=self.config.ema_decay)

    if self.rank == 0:
        self.save_dir = self._get_save_dir()
        self.config.to_yaml(self.save_dir / "train_config.yaml")
        # TensorBoard setup unchanged
    else:
        self.save_dir = Path(self.config.project) / self.config.name
        self.tensorboard_writer = None
```

#### `_train_epoch` (line 439) — EMA unwrap, sampler epoch, rank-0 tqdm

```python
# At the start of the loop body:
if self.is_distributed:
    self.train_loader.sampler.set_epoch(epoch)

# Progress bar only on rank 0:
pbar = tqdm(..., disable=not sys.stderr.isatty() or self.rank != 0)

# EMA update — unwrap the DDP shell:
if self.ema_model is not None:
    raw = self.model.module if self.is_distributed else self.model
    self.ema_model.update(raw)
```

#### `_validate_epoch` (line 550) — rank 0 only

```python
def _validate_epoch(self, epoch):
    if self.rank != 0:
        return None
    # ... existing code unchanged ...
```

Validation uses `self.wrapper_model` and `self.ema_model.ema`; both are already
unwrapped `nn.Module` instances, so no `.module` access is needed here.

#### `_save_checkpoint` (line 619) — rank 0 only, unwrap for state_dict

```python
def _save_checkpoint(self, epoch, loss, val_metrics=None):
    if self.rank != 0:
        return
    raw = self.model.module if self.is_distributed else self.model
    model_to_save = self.ema_model.ema if self.ema_model else raw
    # ... rest of checkpoint logic unchanged, uses model_to_save ...
```

---

### Changes — model-specific trainers

#### YOLOX — `on_mosaic_disable` ([models/yolox/trainer.py:65](libreyolo/models/yolox/trainer.py#L65))

`on_mosaic_disable` is called mid-training, **after** DDP wrapping. It sets an
attribute directly on the model head:

```python
def on_mosaic_disable(self):
    self.train_loader.dataset.close_mosaic()
    raw = self.model.module if self.is_distributed else self.model
    raw.head.use_l1 = True
```

`on_setup` ([line 57](libreyolo/models/yolox/trainer.py#L57)) is called before DDP
wrapping and does not need to change.

#### D-FINE and DEIM — `_train_epoch` gradient clipping

Both trainers override `_train_epoch` and call `self.model.parameters()` for gradient
clipping ([dfine/trainer.py:454,465](libreyolo/models/dfine/trainer.py#L454),
[deim/trainer.py:477,488](libreyolo/models/deim/trainer.py#L477)).

`DDP.parameters()` delegates to the wrapped module, so these calls work unchanged.
The one change needed is `scaler.unscale_` before clipping when AMP is on — this is
already present in both trainers and remains correct.

EMA updates in both custom loops also need the `.module` unwrap:

```python
# In dfine/trainer.py:471 and deim/trainer.py:494
if self.ema_model is not None:
    raw = self.model.module if self.is_distributed else self.model
    self.ema_model.update(raw)
```

The tqdm lines in both overrides should also be gated on `rank == 0` (same pattern as
the base trainer).

#### D-FINE, DEIM, RT-DETR, DEIMv2 — `_setup_optimizer`

All four trainers override `_setup_optimizer` and call `self.model.named_parameters()`
([dfine:159](libreyolo/models/dfine/trainer.py#L159),
[deim:163](libreyolo/models/deim/trainer.py#L163),
[rtdetr:235](libreyolo/models/rtdetr/trainer.py#L235),
[deimv2:179](libreyolo/models/deimv2/trainer.py#L179)).

Because `_setup_optimizer` is called **before** DDP wrapping in `setup()`, these calls
see the raw model and do not need to change.

#### PicoDet — `on_setup`

Accesses `self.model.head` ([picodet/trainer.py:72-74](libreyolo/models/picodet/trainer.py#L72))
to read `num_classes`, `reg_max`, and `strides`. Called before DDP wrapping — no
change needed.

---

### Advantages

- No new dependencies.
- Full control; behaviour is exactly standard PyTorch.
- Slightly faster inter-GPU communication than Accelerate in benchmarks (no wrapper overhead).

### Disadvantages

- Every `_train_epoch` override (D-FINE, DEIM) must independently implement the same
  rank-0 gating, EMA unwrap, and tqdm logic. With 13 model families there is real
  duplication risk.
- Manual `GradScaler` management across all custom training loops.
- `torchrun` is required; single-GPU runs must still use `torchrun --nproc_per_node=1`
  or add a guard to skip `init_process_group`.
- Subtle ordering constraint: DDP wrapping must happen after optimizer setup but before
  EMA creation, making `setup()` fragile to reordering.

---

## Approach 2 — HuggingFace Accelerate

### How it works

Accelerate wraps `model`, `optimizer`, and `dataloader` through a single
`accelerator.prepare()` call. It handles DDP wrapping, `DistributedSampler`,
mixed-precision scaler, and rank-0 gating behind a stable API. The training loop
itself changes minimally.

### Install

```bash
pip install accelerate
accelerate config   # one-time wizard that writes ~/.cache/huggingface/accelerate/default_config.yaml
```

### Launch

```bash
accelerate launch train.py ...
```

Or equivalently, `torchrun` still works — Accelerate reads `LOCAL_RANK` /
`WORLD_SIZE` from the environment just like raw DDP.

---

### Changes — `libreyolo/training/trainer.py`

#### `__init__` — create the Accelerator

```python
from accelerate import Accelerator

mp = "fp16" if self.config.amp else "no"
self.accelerator = Accelerator(mixed_precision=mp)
self.rank = self.accelerator.process_index
self.is_distributed = self.accelerator.num_processes > 1
```

#### `_setup_device` (line 156) — delegate to accelerator

```python
def _setup_device(self) -> torch.device:
    return self.accelerator.device
```

The `NotImplementedError` and all the `device_str` parsing are replaced entirely.
The accelerator resolves the device from `LOCAL_RANK` automatically.

#### `effective_lr` property (line 88) — scale by world size

Same correction as DDP:

```python
@property
def effective_lr(self) -> float:
    return self.config.lr0 * self.config.batch * self.accelerator.num_processes / 64
```

#### `setup` (line 339) — remove GradScaler, call `prepare`, gate I/O

```python
def setup(self):
    if self._is_setup:
        return

    self.model.to(self.device)
    self.on_setup()
    self._setup_data()
    self.optimizer = self._setup_optimizer()

    # Replace the manual GradScaler entirely — Accelerate owns AMP
    self.scaler = None

    # Single call wraps model, optimizer, and dataloader
    self.model, self.optimizer, self.train_loader = self.accelerator.prepare(
        self.model, self.optimizer, self.train_loader
    )
    # prepare() adds DistributedSampler automatically

    if self.config.ema:
        raw = self.accelerator.unwrap_model(self.model)
        self.ema_model = ModelEMA(raw, decay=self.config.ema_decay)

    if self.accelerator.is_main_process:
        self.save_dir = self._get_save_dir()
        self.config.to_yaml(self.save_dir / "train_config.yaml")
        # TensorBoard setup unchanged
    else:
        self.save_dir = Path(self.config.project) / self.config.name
        self.tensorboard_writer = None
```

#### `_train_epoch` (line 439) — replace backward, gate tqdm, unwrap EMA

```python
# Progress bar:
pbar = tqdm(..., disable=not sys.stderr.isatty() or not self.accelerator.is_main_process)

# Replace both the AMP and non-AMP backward paths with one call:
self.accelerator.backward(loss)

# Remove: self.scaler.scale(loss).backward()
# Remove: self.scaler.step(self.optimizer)
# Remove: self.scaler.update()
# Keep:   self.optimizer.step() — Accelerate handles unscaling internally

# EMA:
if self.ema_model is not None:
    raw = self.accelerator.unwrap_model(self.model)
    self.ema_model.update(raw)
```

The `if self.scaler is not None` / `else` branching collapses into a single code path.

#### `_validate_epoch` (line 550) — main process only

```python
def _validate_epoch(self, epoch):
    if not self.accelerator.is_main_process:
        return None
    # ... existing code unchanged ...
```

#### `_save_checkpoint` (line 619) — main process only, unwrap

```python
def _save_checkpoint(self, epoch, loss, val_metrics=None):
    if not self.accelerator.is_main_process:
        return
    raw = self.accelerator.unwrap_model(self.model)
    model_to_save = self.ema_model.ema if self.ema_model else raw
    # ... rest unchanged ...
```

---

### Changes — model-specific trainers

#### YOLOX — `on_mosaic_disable` ([models/yolox/trainer.py:65](libreyolo/models/yolox/trainer.py#L65))

Same as DDP; `on_mosaic_disable` fires after `prepare()`:

```python
def on_mosaic_disable(self):
    self.train_loader.dataset.close_mosaic()
    self.accelerator.unwrap_model(self.model).head.use_l1 = True
```

#### D-FINE and DEIM — `_train_epoch`

Both overrides contain the manual AMP split. Replace with Accelerate's backward and
remove the `scaler` branches:

```python
# Before (AMP path):
self.scaler.scale(loss).backward()
if is_opt_step:
    if clip_max_norm > 0:
        self.scaler.unscale_(self.optimizer)
        torch.nn.utils.clip_grad_norm_(self.model.parameters(), clip_max_norm)
    self.scaler.step(self.optimizer)
    self.scaler.update()

# Before (non-AMP path):
loss.backward()
if is_opt_step:
    if clip_max_norm > 0:
        torch.nn.utils.clip_grad_norm_(self.model.parameters(), clip_max_norm)
    self.optimizer.step()

# After (both paths collapsed):
self.accelerator.backward(loss)
if is_opt_step:
    if clip_max_norm > 0:
        self.accelerator.clip_grad_norm_(self.model.parameters(), clip_max_norm)
    self.optimizer.step()
```

`accelerator.clip_grad_norm_` handles the unscaling step internally when AMP is active.

EMA unwrap in both loops:

```python
if self.ema_model is not None:
    self.ema_model.update(self.accelerator.unwrap_model(self.model))
```

Tqdm gating: same as base trainer — gate on `self.accelerator.is_main_process`.

#### D-FINE, DEIM, RT-DETR, DEIMv2 — `_setup_optimizer`

Called before `prepare()`, see the raw model. No changes needed.

#### PicoDet — `on_setup`

Called before `prepare()`. No changes needed.

---

### Advantages

- The entire `GradScaler` management (two code paths in every `_train_epoch`)
  collapses to `accelerator.backward(loss)` — one line regardless of AMP.
- `is_main_process` and `unwrap_model` are consistent across all training loops;
  the same pattern applies to every `_train_epoch` override without thinking about
  DDP vs single-GPU.
- Single-GPU runs work without any code change — the accelerator is a no-op when
  `num_processes == 1`.
- No ordering constraint between optimizer setup and model wrapping — `prepare()`
  handles everything.

### Disadvantages

- Adds `accelerate` as a runtime dependency.
- The `accelerator` instance must be threaded through the entire trainer hierarchy.
  Subclasses that override `_train_epoch` need access to it, which means it must live
  on `self` and be constructed early.
- Behaviour in edge cases (e.g. gradient accumulation with AMP across DDP) is
  controlled by Accelerate's internals; debugging requires understanding both the
  Accelerate layer and the underlying PyTorch primitives.
- `accelerate config` requires a one-time interactive setup step, which adds friction
  in CI or containerised environments (can be bypassed with a config file).

---

## Summary comparison

| | Raw DDP | Accelerate |
|---|---|---|
| New dependency | None | `accelerate` |
| Launch command | `torchrun` | `accelerate launch` (or `torchrun`) |
| AMP handling | Manual `GradScaler` in every loop | `accelerator.backward(loss)` |
| Single-GPU compat | Requires guard around `init_process_group` | Transparent — no-op |
| `_train_epoch` overrides (D-FINE, DEIM) | Each needs its own rank-gating + EMA unwrap | Same pattern, same API |
| `.module` unwrapping | `model.module` — must track DDP state manually | `accelerator.unwrap_model(model)` — works on DDP and plain models |
| Setup ordering | DDP wrap must occur after optimizer, before EMA | `prepare()` handles ordering |
| Debugging | Pure PyTorch — no abstraction layer | One extra layer to understand |

For this codebase the Accelerate path is lower risk: the two custom `_train_epoch`
overrides (D-FINE and DEIM) are the most error-prone part of the DDP approach, and
Accelerate's uniform API removes most of that per-file variation.
